### PR TITLE
Add support for testing with travis

### PR DESCRIPTION
**Pull request by wpoely86**
It's dead simple and it works!

### CODE REVIEW

**Comment by wpoely86** [comment]
For example: https://travis-ci.org/wpoely86/logstash-patterns

**Comment by stdweird** [comment]
how is this easier than adding it to jenkins? also, what is the realtion between what you are testing and what you are deploying?

**Comment by wpoely86** [comment]
The fact that it's all controlled by a single file that is part of the repository is a major advantage IMHO. No dependency on external config files and a way to track changes.

For this repo, it's currently not testing deployment as the package only contains a bunch of files without tests. But in general nothing is stopping you from deploying first and then testing.

**Comment by boegel** [comment]
@stdweird this is 15 lines of YAML; on Jenkins, the test config is all over the place (in Jenkins itself, but also on the filesystem where dependent packages are 'just there', etc.

I'm working on switching to Travis for EasyBuild too, where even more advantages come forward: more resources to run the tests (and hence less pressure on Jenkins which benefits stuff we can't easily test outside of Jenkins), and making it trivial for others to test their branches with exactly the same test config as used centrally (which means they can make sure their contribution is OK before issuing a PR).

We should discuss this again during team meeting imho (although my mind is already made up w.r.t. EB, tbh).

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+language: python
+python: 2.7
```
Comment (boegel) [inline .travis.yml]: also test 2.6?

```
python:
  - 2.6
  - 2.7
```

Comment (wpoely86) [inline .travis.yml]: why? It's only the tests itself that is actual python. All the rest is just grok patterns.

Comment (boegel) [inline .travis.yml]: Ah, ok, doesn't make sense then indeed. :)

**Comment by stdweird** [comment]
@boegel this is not yaml, this is funny formatted bash (look at the install and script section).
wrt the dependencies, you can also pull in dependencies on jenkins (e.g. `python setup.py test` will do this if the dependencies are well defined); or make them part of the test, like travis does this.
the trouble to add this to jenkins is minimal (altough i'll admit that the jenkins config sucks). not sure if this is worth any effort (or were you thinking of not moving all tests to travis?)

**Comment by boegel** [comment]
@stdweird You can pull in certain deps on Jenkins too, sure, but for some reason we haven't (even though the effort is minimal). And even that's not 100% true, since we want to test things on the `develop` version of the deps, which may not have a matching release yet. Also, testing on top of both Python 2.6 and 2.7 is a f*cking mess on Jenkins, while it's trivial with Travis (not relevant here, but very much for most of our other stuff). +1 for Travis.

W.r.t. syntax: I don't really care whether you call it YAML or funny formatted bash, the point is that it's a single (readable) config that specifies *everything*, and that it's under version control in the repository itself, without getting in the way. +1 for Travis.

I'm not advocating to move to Travis for everything, I wasn't even planning on pushing it for other things besides EasyBuild (except *maybe* vsc-base). For simple projects/test configs, the benefit is limited. For something complex enough as EasyBuild, where we run the tests in multiple different configurations (Lmod vs env mods, Tcl vs Lua syntax of module files, etc.) and where we need to test different branches (master, develop and PRs), it's a great simplification (with added extra benefits). Everything is in one place, the same test config is used for different branches without having to do anything. +1 for Travis.

For example, right now we try to keep the test config for `easybuild-framework` in sync across three different Jenkins jobs (for master, develop and PRs), and have our own custom (manual) installations of Python 2.6 and 2.7 sitting in the home dir of `jenkins`. Replicating that setup would require a lot of effort, in case it's ever needed, because it's spread out all over the place. Moreover, this approach has bitten us in the ass a couple of times already: problems with Jenkins made one of the merge sprints fail a while ago, and I ran into bugs during the release of EasyBuild 2.7.0 that we missed because of our crappy test setup. Frankly, I'm sick of it, and I believe Travis would be a major help here (more resources, consistent test config). +1 for Travis.

If you see any reasons why we should stick to Jenkins (for EasyBuild), please share. I don't see any now that I spent some time making myself more familiar with Travis, to be honest.

**Comment by wpoely86** [comment]
Ping? 

Do we merge or not? It works and Jenkins currently doesn't test this.

**Comment by boegel** [comment]
I'm all for it. @stdweird?

**Comment by stdweird** [comment]
i'm not :smile: 
just do the wget once in jenkins and get over it

**Comment by boegel** [comment]
@stdweird I guess your main concern is not to start scattering repos across Jenkins and Travis? Fine.

@JensTimmerman What's the best way to set up a new job in Jenkins to test a repo that wasn't tested before (incl. tracking what needs to be put in place as deps for that job)?

**Comment by JensTimmerman** [comment]
Using the jenkins pipeline plugin (which will be default in jenkins2, which is comming out soon)
you can create a groovy script that specifies what needs to run, e.g. setup a few dependencies, set the path, and run `python setup.py test`
see e.g. https://github.com/JensTimmerman/vsc-accountpage/blob/test_branch/Jenkinsfile

**Comment by wpoely86** [comment]
@JensTimmerman Nice! The plugin is installed on our current Jenkins servers?

**Comment by boegel** [comment]
If the `.travis.yml` file was weirdly formatted bash, this is even worse (strings galore, need to escape `$`, etc.).

Just sayin'. ;-)

**Comment by stdweird** [comment]
yep, worse indeed. don't understand why we don't create a bash script and run that.

**Comment by JensTimmerman** [comment]
wel, this is an example, and I don't know any groovy yet, so I hacked this toghether as a proof of concept, Sure a bashscript will shorten this string galore down.

The groovy will play a part into explaining to jenkins what parts or things can be run in parallel, e.g. 2 times the same script with a different python etc.
https://github.com/jenkinsci/pipeline-plugin/blob/master/TUTORIAL.md

an example of a very complicated build procedure: 
(run tests, deploy to qa, manual input from someone to confirm deploy to production)
https://github.com/jenkinsci/workflow-aggregator-plugin/blob/master/demo/repo/Jenkinsfile


@wpoely86 it's installed on jenkins0, if you want to start testing/playing around with it there, see the tutorial linked above on info about it, and the test-branch-jens job as an example on how to easily tell jenkins how to get this Jenkinsfile 
Jenkins2 was released today, so the plugins got a big version bump, but everything seems normal ;-)

**Comment by wpoely86** [comment]
Superseded by #40

**Pull request closed without merging**